src/Lykke.Cqrs.MessageCancellation/Services/MessageCancellationService.cs

```diff
@@ -0,0 +1,87 @@
+﻿using Lykke.Cqrs.MessageCancellation.Services.Interfaces;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Lykke.Cqrs.MessageCancellation.Services
+{
+    public class MessageCancellationService : IMessageCancellationService
+    {
+        private ReaderWriterLockSlim _readWriterLockSlim;
+        private HashSet<string> _operationsToCancel;
```

Comment:
In case we have multiple instances of the same app, message cancellation call via api would be processed by only one instance. If target message is processed by  another instance, will it end up at some point in first instance that knows about cancellation request? If no, then I suggest to also add Redis cache for cancellation requested ids. If redis is down then we can fallack to local HashSet.

Reply:
In the task it is clearly stated to make the functionality stateful. 
That feature has a potential to become distributed in the future.
But now it is needed in the way it is implemented.